examples/quad/main.rs

```diff
@@ -137,663 +138,751 @@ fn main() {
-
-    // copy image data into staging buffer
-    unsafe {
-        let mut data = device
-            .acquire_mapping_writer::<u8>(&image_upload_memory, 0 .. image_mem_reqs.size)
-            .unwrap();
-        for y in 0 .. height as usize {
-            let row = &(*img)
-                [y * (width as usize) * image_stride .. (y + 1) * (width as usize) * image_stride];
-            let dest_base = y * row_pitch as usize;
-            data[dest_base .. dest_base + row.len()].copy_from_slice(row);
+        match event {
+            winit::event::Event::WindowEvent { event, .. } => {
+                match event {
+                    winit::event::WindowEvent::CloseRequested => *control_flow = winit::event_loop::ControlFlow::Exit,
+                    winit::event::WindowEvent::KeyboardInput {
+                        input:
+                            winit::event::KeyboardInput {
+                                virtual_keycode: Some(winit::event::VirtualKeyCode::Escape),
+                                ..
+                            },
+                        ..
+                    } => *control_flow = winit::event_loop::ControlFlow::Exit,
+                    winit::event::WindowEvent::Resized(dims) => {
+                        println!("resized to {:?}", dims);
+                        let dimensions = Extent2D {
+                            width: dims.width as u32,
+                            height: dims.height as u32,
+                        };
+                        renderer.resize_swapchain(dimensions);
+                    }
+                    _ => { }
+                }
+            }
+            winit::event::Event::EventsCleared => {
+                renderer.render();
+            }
+            _ => { }
         }
-        device.release_mapping_writer(data).unwrap();
-    }
+    });
+}
 
-    let mut image_logo = unsafe {
-        device.create_image(
-            kind,
-            1,
-            ColorFormat::SELF,
-            i::Tiling::Optimal,
-            i::Usage::TRANSFER_DST | i::Usage::SAMPLED,
-            i::ViewCapabilities::empty(),
-        )
-    }
-    .unwrap();
-    let image_req = unsafe { device.get_image_requirements(&image_logo) };
-
-    let device_type = memory_types
-        .iter()
-        .enumerate()
-        .position(|(id, memory_type)| {
-            image_req.type_mask & (1 << id) != 0
-                && memory_type.properties.contains(m::Properties::DEVICE_LOCAL)
-        })
-        .unwrap()
-        .into();
-    let image_memory = unsafe { device.allocate_memory(device_type, image_req.size) }.unwrap();
-
-    unsafe { device.bind_image_memory(&image_memory, 0, &mut image_logo) }.unwrap();
-    let image_srv = unsafe {
-        device.create_image_view(
-            &image_logo,
-            i::ViewKind::D2,
-            ColorFormat::SELF,
-            Swizzle::NO,
-            COLOR_RANGE.clone(),
-        )
-    }
-    .unwrap();
+struct Renderer<B: hal::Backend> {
```

Comment:
do we have to make it structured here? I think having a very simple example that just does all the things in one function is a good point to start with, and it's less code than trying to structurize everything. We have colour-uniform to cover that use case

Reply:
Because winit now calls us I think storing resources in a struct is now worth it:
*   Much easier to understand as the winit events loop code is cleanly separated from the renderer code
*   Putting the cleanup code in the Renderer::drop is less scary, in terms of never cleaning up or cleaning up more than once.
*   All resources that are not used in the event loop are never dropped. When resource creation is done in a `Renderer::new`, then all resources created during the function either go into the Renderer struct or are dropped at the end of `Renderer::new`. Without that external function there are some gfx structs that are not used directly in the render loop so they do not get dropped, I dont know if this is an issue or not.

I have however created a separate branch here without the Renderer struct: https://github.com/rukai/gfx/commit/7247ed8e46cbbbe57a373f02b80f129ff1d69276
Let me know if you would still prefer me to replace this branch with that one.

Reply:
Thank you! Would it be possible to write this without `ManuallyDrop`?
If the structured renderer is the way to go for all of the examples, and there is enough winit-related boilerplate, we may move it into "main_loop.rs" or something, shared by all the window-related examples.

Reply:
Manually drop is required because the move closure only gives us an &mut T